Improve R package creation

glennpj opened the pull request:
This PR adds the `list_url` attribute to CRAN R packages when using
`spack create`. It also adds the `git` attribute to R Bioconductor
packages upon creation.

glennpj committed:
Improve R package creation

This PR adds the `list_url` attribute to CRAN R packages when using
`spack create`. It also adds the `git` attribute to R Bioconductor
packages upon creation.

glennpj commented:
It would probably be good to add a `cran` attribute, similar to the `pypi` attribute for python packages. That is something that could be done with the next round of R package updates, so the attribute can be assigned to existing packages.

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -328,14 +328,28 @@ def configure_args(self):
+                self.url_line = '    url      = "{0}"\n'\
+                    '    list_url = "{1}"'.format(url, list_url)
+            elif re.search(r'.*bioconductor.org', url):
+                git = "https://git.bioconductor.org/packages/{0}".format(r_name)
```
IMO It is not appropriate or desirable to hard-code URL paths into core Spack commands.  

If these are necessary for R packages then I suggest they be added to `RPackage` and accessed here.

tldahlgren commented on lib/spack/spack/cmd/create.py:
```diff
@@ -328,14 +328,28 @@ def configure_args(self):
+                self.url_line = '    url      = "{0}"\n'\
+                    '    list_url = "{1}"'.format(url, list_url)
+            elif re.search(r'.*bioconductor.org', url):
+                git = "https://git.bioconductor.org/packages/{0}".format(r_name)
```
IMO Hard-coded URLs are not appropriate in the general Spack commands.

tldahlgren replied:
I'm not familiar with `R`.  Are these the standard or canonical URLs for such packages?

glennpj replied:
Yes, all CRAN R packages are located at https://cloud.r-project.org and all Bioconductor R packages are located at https://bioconductor.org.

adamjstewart commented:
I'm fine with this change, although if we add a `cran` attribute we'll first need to undo this change. Thoughts on adding `cran` instead of this PR?

glennpj commented:
> Thoughts on adding `cran` instead of this PR?

I can do that, in fact I had started down that path first. I would want to set that for all of the CRAN packages, so it would wind up being a largish PR, although I guess that is not strictly necessary to do now. The other thing is that does not address needing to set the `git` attribute for bioconductor packages. The bioconductor packages are what led me to the solution in this PR. I was trying to be "consistent" in how the two types were handled. The bioconductor packages are harder to deal with as really the checksum should be changed to the git commit but I am not sure how to tackle that during package creation. Setting the git repo may be as far as this can go in the creation step.

For context, it is likely that a bioconductor package is created from the url of the current version. The problem is that the url will be of the form, https://bioconductor.org/packages/release/bioc/src/contrib/package_X.YY.Z.tar.gz, meaning that when the next version of bioconductor is released, the spack packages using that URL form will no longer be able to fetch the package. The package version numbering and the bioconductor version numbering are not related in a predictable way across packages. The URL could be changed to replace 'release' in the url to the actual bioconductor release but that is not scalable. Using git and branches based on the bioconductor release make it easier to manage updates.

An idea would be to add a `bioc` attribute, similar to `cran`, but that sets the git repo for bioconductor packages instead of the url. That would mean that a new bioconductor package created with `spack create` would have the url used to generate the package and a `bioc` attribute. That would eventually need to be changed for the reason mentioned above, either during PR review, or when the spack bioconductor packages are updated to the next bioconductor release. Would that be okay, or would that be too confusing, given that the analogous `pypi` and `cran` attributes set the url?

glennpj commented:
I will also add that I was planning on doing some documentation edits to https://spack.readthedocs.io/en/latest/build_systems/rpackage.html to address the difficulties posed with bioconductor packages.

adamjstewart commented:
It's fine for `bioc` to set `git` instead of `url`. I wouldn't use `bioc` unless it makes things easier. For example, the only reason I added `pypi` was so it could automatically set `homepage`, `url`, and `list_url`. Same for `cran`.

glennpj committed:
Switch over to using cran/bioc attributes

The cran/bioc entries are set to have the '=' line up with homepage
entry, but homepage does not need to exist in the package file. If it
does not, that could affect the alignment.

glennpj commented:
I just uploaded a commit to use `cran` and `bioc` attributes. I feel that switching to those requires a documentation update, more so than what I had previously, so I will keep this as a draft PR until I have the documentation PR ready.

glennpj committed:
Do not have to split bioc

glennpj committed:
Edit R package documentation

Explain Bioconductor packages and add `cran` and `bioc` attributes.

adamjstewart commented on lib/spack/docs/build_systems/rpackage.rst:
```diff
@@ -124,6 +126,67 @@ If you only specify the URL for the latest release, your package will
+Bioconductor packages
+^^^^^^^^^^^^^^^^^^^^^
+
+Bioconductor packages are setup in a similar way to CRAN packages, but there
```
```suggestion
Bioconductor packages are set up in a similar way to CRAN packages, but there
```

adamjstewart commented on lib/spack/docs/build_systems/rpackage.rst:
```diff
@@ -124,6 +126,67 @@ If you only specify the URL for the latest release, your package will
+Bioconductor itself is versioned and released. This scheme, using the
+Bioconductor package installer, allows further specification of the minimum
+version of R as well as further restrictions on the dependencies between
+packages than what is possible with the native R packaging system. Spack can
+not replicate these extra features and thus Bioconductor packages in Spack need
+to be managed as a group during updates in order to maintain package
+consistency with Bioconductor itself.
```
We could simply create a bioconductor `BundlePackage` that uses the versioning scheme and package versions that Bioconductor uses. See the `xsdk` package for an example.

glennpj replied:
I have thought of this, and think I suggested it in the past. I am on the fence about it though. First off, a bundle package would not address the fact that bioconductor packages would need to be updated as a group. It would not help with package creation, PR review, or updates. In fact, it would add one more level as new and updated packages would need to be added to the bundle. All of the bioconductor packages would need to be added with version specs to avoid the possibility of some packages picking up a newer version of a dependency if an older bioconductor bundle is installed. The package file would be quite large and ever growing. Finally, I am not sure that people really want to install all of the spack bioconductor packages, although that may be viewed as reasonable if the bundle package existed.

glennpj replied:
I have thought of this, and think I suggested it in the past. I am on the fence about it though. First off, a bundle package would not address the fact that bioconductor packages would need to be updated as a group. It would not help with package creation, PR review, or updates. In fact, it would add one more level as new and updated packages would need to be added to the bundle. All of the bioconductor packages would need to be added with version specs to avoid the possibility of some packages picking up a newer version of a dependency if an older bioconductor bundle is installed. The package file would be quite large and ever growing. Finally, I am not sure that people really want to install all of the spack bioconductor packages, although that may be viewed as reasonable if the bundle package existed.

adamjstewart commented on lib/spack/docs/build_systems/rpackage.rst:
```diff
@@ -124,6 +126,67 @@ If you only specify the URL for the latest release, your package will
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Much like the ``pypi`` attribute for python packages, due to the fact that R
+packages are obtained from specific repositories, it is possible to setup shortcut
```
```suggestion
packages are obtained from specific repositories, it is possible to set up shortcut
```

adamjstewart commented on lib/spack/docs/build_systems/rpackage.rst:
```diff
@@ -124,6 +126,67 @@ If you only specify the URL for the latest release, your package will
+
+.. code-block:: python
+
+   cran = 'caret/caret_6.0-86.tar.gz'
```
The reason that the `pypi` attribute has this structure is because:

1. The package name may not always match the tarball name (e.g. `Foo_Bar/foo-bar-1.2.3.tar.gz`)
2. The package may not always use tar (e.g. `foo-bar/foo-bar-1.2.3.zip`)

Is this also the case for CRAN? If not, perhaps we could make this simpler and users would only need to set `cran = 'caret'`.

glennpj replied:
I could be missing something but I think the form with the tarfile is needed for setting the `url` attribute. The `bioc` attribute only needs to set the `git` attribute so it uses the simpler form.

adamjstewart replied:
Yes, but if a tarball is always available, you could simply set `cran = 'caret'` and set the url to a fake version like `{cran}_1.2.3.tar.gz`.

glennpj replied:
There are some differences in the way r- packages specify versions though. Some variants:
```
{cran}_1.2.3
{cran}_1.2-3
{cran}_1-2
{cran}_1.2
{cran}_1.2.3-1
{cran}_1.02
{cran}_1.2.3.4
{cran}_1.2-3.4
{cran}_1.2.3.4.5
```
There are also some that use date format versions. That does not seem like it could be faked.

adamjstewart replied:
But that part doesn't matter, as long as Spack detects the version in the URL and replaces it.

glennpj replied:
Okay; I underestimated Spack's power in version processing then. It looks like all of the CRAN packages are `tar.gz` files so I will push out a change later today.

Thanks for pointing that out.

adamjstewart replied:
Just need to update this line of the docs now.

glennpj replied:
Sorry, I missed that. Should the url example be left as-is or changed to reflect the stand-in version?
```
.. code-block:: python

   cran = 'caret'

is equivalent to:

.. code-block:: python

   homepage = 'https://cloud.r-project.org/package=caret'
   url      = 'https://cloud.r-project.org/src/contrib/caret_6.0-86.tar.gz'
   list_url = 'https://cloud.r-project.org/src/contrib/Archive/caret'
```

glennpj replied:
I think it makes sense to leave it as-is. That makes it consistent with the format that is presented with `spack info`.

glennpj committed:
Update lib/spack/docs/build_systems/rpackage.rst

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Update lib/spack/docs/build_systems/rpackage.rst

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

glennpj committed:
Simplify the cran attribute

The version can be faked so that the cran attribute is simply equal to
the CRAN package name.

adamjstewart commented on lib/spack/spack/build_systems/r.py:
```diff
@@ -33,6 +41,34 @@ class RPackage(PackageBase):
+        if self.cran:
+            return (
+                'https://cloud.r-project.org/src/contrib/'
+                + self.cran + '_1.2.3.tar.gz'
```
Does `self.versions` contain a dict of versions we can access? It doesn't really matter if we use a fake version, but `spack url summary` uses these numbers to compute our url parsing accuracy.

glennpj replied:
I am not following what you mean.

glennpj replied:
Do you mean something like this:
```python
    @property
    def url(self):
        if self.cran:
            for version in self.versions:
                return (
                    'https://cloud.r-project.org/src/contrib/'
                    + self.cran + '_' + str(version) + '.tar.gz'
                )
```

adamjstewart replied:
Something like:
```python
    @property
    def url(self):
        if self.cran:
            return (
                'https://cloud.r-project.org/src/contrib/'
                + self.cran + '_' + str(self.versions[0]) + '.tar.gz'
            )
```
Just so it isn't an arbitrary fake version, but an actual version from the package.

glennpj replied:
Okay; I understand. I thought you wanted the whole dict of versions, which is why I was confused.

glennpj committed:
Edit the docs to reflect new `cran` attribute format

glennpj committed:
Use the first element of self.versions() for url

adamjstewart commented on lib/spack/spack/build_systems/r.py:
```diff
@@ -33,6 +41,34 @@ class RPackage(PackageBase):
+        if self.cran:
+            return (
+                'https://cloud.r-project.org/src/contrib/'
+                + self.cran + '_' + str(list(self.versions)[0]) + '.tar.gz'
```
Does this work during `spack create`? I'm not sure if `self.url` is accessed before we officially have a list of versions.

glennpj replied:
I tested with a CRAN package that is not in spack already using the following commands:

- `spack create https://cran.r-project.org/src/contrib/BinGSD_0.0.1.tar.gz`
- `spack info r-bingsd`
- `spack stage r-bingsd`

All of those worked fine. I am not sure how to verify for `spack url summary`.

glennpj replied:
The `BinGSD` example that I grabbed is a new CRAN package and does not have an Archive directory yet. So, I also modified `r-caret` to use the `cran` attribute and tested 

- `spack versions r-caret`
- `spack checksum r-caret <<<2`

Those worked fine.

adamjstewart replied:
> I am not sure how to verify for `spack url summary`

Basically you can run `spack url summary` on develop and on your branch and they should be similar/the same for version detection.

glennpj replied:
I rebased my branch and I get the same exact output for both branches.
